WIP: Uptime tacking

brett060102 opened the pull request:
## What does this PR change?

Adds collect and reporting of uptime data to the SCC for SUSE  branded OSes

## GUI diff

No difference.

Before:

After:

- [ ] **DONE**

## Documentation
- No documentation needed: **add explanation. This can't be used if there is a GUI diff**
- No documentation needed: only internal and user invisible changes
- Documentation issue was created: [Link for SUSE Manager contributors](https://github.com/SUSE/spacewalk/issues/new?template=ISSUE_TEMPLATE_DOCUMENTATION.md&labels=documentation&projects=SUSE/spacewalk/31), [Link for community contributors](https://github.com/uyuni-project/uyuni-docs/issues/new).
- API documentation added: please review the Wiki page [Writing Documentation for the API](https://github.com/uyuni-project/uyuni/wiki/Writing-documentation-for-the-API) if you have any changes to API documentation.
- (OPTIONAL) [Documentation PR](https://github.com/uyuni-project/uyuni-docs/pulls)

- [ ] **DONE**

## Test coverage
- No tests: **add explanation**
- No tests: already covered
- Unit tests were added
- Cucumber tests were added

- [ ] **DONE**

## Links

Issue(s): #
Port(s): # **add downstream PR(s), if any**

- [ ] **DONE**

## Changelogs

Make sure the changelogs entries you are adding are compliant with https://github.com/uyuni-project/uyuni/wiki/Contributing#changelogs and https://github.com/uyuni-project/uyuni/wiki/Contributing#uyuni-projectuyuni-repository

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

# Before you merge

Check [How to branch and merge properly](https://github.com/uyuni-project/uyuni/wiki/How-to-branch-and-merge-properly)!

mcalmer commented on java/code/src/com/suse/manager/webui/utils/salt/custom/UptimeTrackerRequestData.java:
```diff
@@ -0,0 +1,60 @@
+     * Gets all info returned as Map for inserting into the report entry
+     * @return returns data as Map.
+     */
+    public Map<String, Object> asMap() {
+        Map<String, Object> out = new HashMap<>();
+        getUptimeTrackeReport()
+                .map(StateApplyResult::getChanges)
+                .ifPresent(c -> {
+                    if (c.getRetcode() == 0) {
+                        out.put("dump-uptime-info", c.getStdout());
+                    }
+                });
+
+        return out;
+    }
```
This is not strictly needed. I added it to CoCoAttestation class because we plan to enhance the state by new checks and we want to copy the value 1:1 into the database.
You have 1 specified value and you can call `getUptimeTrackerReport()` also directly at the place where you call it and write the value into the database.
If NULL is a valid value in the database for this column, you also do not need the "Optional" handling.

brett060102 replied:
Changes pushed. I hope I got that right.

mcalmer commented on susemanager-utils/susemanager-sls/salt/uptime-tracker/suse-upstream-tracker.sls:
```diff
@@ -0,0 +1,35 @@
+{%- if not salt['pillar.get']('susemanager:distupgrade:dryrun', False) %}
```
I think you do not add it to the distupgrade state.
I think you do not need this "if" condition.

brett060102 replied:
Thank you. I am going to pull everything but dump-uptime-info from the file. The other items were really just to make my testing easier. Ultimately installation of the package and enabling the collection will be done by the install of the base OS.

brett060102 replied:
Cleaned unneeded code from the file. I left os_family Suse check, because we only need to check for the file on SUSE based systems. I did not know if having the checked kept the server from having the action run on the client or not. If the action is run on the client anyway, then that check can also be removed.

brett060102 committed:
Add support for suse uptime tracker

Squashed commit of the following:
commit aa733d712e1cef3b8153ee5584308b9fad99825e
Author: sampsone <esampson@suse.com>
Date:   Tue Jul 30 15:10:49 2024 -0500

    add UptimeTracker to SaltUtils

commit 096773caed0cae9878bb10e2cdcdc2887498f87a
Author: sampsone <esampson@suse.com>
Date:   Mon Jul 29 15:48:38 2024 -0500

    Add schema upgrade for uptimetracker into suseInternalState

commit 27dd4ce765e42eddf8326efdbed462edfde0527f
Author: sampsone <esampson@suse.com>
Date:   Thu Jul 25 16:17:35 2024 -0500

    Add uptimetracker as a recurring action

commit 6638e9521878abeab13c2dfc9fd6163f4ab24e86
Author: sampsone <esampson@suse.com>
Date:   Tue Jul 9 15:21:04 2024 -0500

    Code cleanup

    1: wrap lines > 130 chars
    2: change dump-uptime-info to dump_uptime_info
    3: make SerializedName for dump_uptime_info match entry for mgr_snpguest_report

commit 444b5be320e2d20ad34afd1621bebfab45443284
Author: sampsone <esampson@suse.com>
Date:   Tue Jul 2 11:09:40 2024 -0500

    fix change uptime_data to uptimeData

commit 8933c568ba244c06833b9da04fbe4cd9a79b2a34
Author: sampsone <esampson@suse.com>
Date:   Tue Jul 2 10:06:51 2024 -0500

    Fix dropped space in SELECT line.

commit 74a6067760cf04eebfedda2b0004990d163f2371
Author: sampsone <esampson@suse.com>
Date:   Thu Jun 27 12:33:28 2024 -0500

    Added change log entry for uptime tracking

commit 7c58155555b6ab16fdade619bf29c0d2feb17383
Merge: 0633dc9a646 78efba659a2
Author: sampsone <esampson@suse.com>
Date:   Thu Jun 27 10:59:32 2024 -0500

    Add support for suse uptime tracker

    The suse uptime tracker package records ever hour a system is up
    as bit mask. This information if present, needs to be collect by suma and
    reported to SCC.

commit 78efba659a29e09bd1bbf8ff81e7ebcb96624596
Author: sampsone <esampson@suse.com>
Date:   Thu Jun 27 09:52:18 2024 -0500

    Update testing for Uptime data

commit 1e243fcfd7e3c967d2e354870f6b2f0dad21311b
Merge: 8800c4a3c26 133f14bad41
Author: sampsone <esampson@suse.com>
Date:   Tue Jun 25 11:02:19 2024 -0500

    Add support for suse uptime tracker

    The suse uptime tracker package records ever hour a system is up
    as bit mask. This information if present, needs to be collect by suma and
    reported to SCC.

    Merge branch 'uptime-tacking' into resync

commit 133f14bad414c2796d094a05248ec913048af699
Author: sampsone <esampson@suse.com>
Date:   Tue Jun 11 19:07:06 2024 -0500

    Address errors found in testing

commit 4f3e5d57a015009a50458b2ad0bfecfd19d5e055
Merge: cb6f2f75597 f4bca0c86f3
Author: sampsone <esampson@suse.com>
Date:   Tue Jun 11 10:52:07 2024 -0500

    Merge branch 'uptime-tacking' into synce-with-up

commit f4bca0c86f31b198ec64591271351af0fe2f3dee
Author: sampsone <esampson@suse.com>
Date:   Tue Jun 11 10:36:23 2024 -0500

    More updates to support uptime tracking

    need to addsupport to get uptime record from DB and send it to SCC as
    online_at.

commit cc93ab27aaa08520aaff2d35ce7ee2dc1c0062df
Author: sampsone <esampson@suse.com>
Date:   Thu May 30 15:00:22 2024 -0500

    update suse-uptime-tracker.sls

    1: onlt execute if file exists nd is not empty
    2: reformat text as json list

commit 9bb26f07e46e519142bbec7622abcf5e28f78797
Author: sampsone <esampson@suse.com>
Date:   Wed May 29 17:50:19 2024 -0500

    fix typo suse-upstream-tracker.sls to suse-uptime-tracker.sls

commit 0acb0d3618aea6be19a1b52af0f3050e71fe2a4b
Author: sampsone <esampson@suse.com>
Date:   Wed May 22 14:16:55 2024 -0500

    Update in response to review comments.

    UptimeTrackerRequestData.java:
       no need to create a map
       since null is OK can just return uptimeTrackerResult

    suse-upstream-tracker.sls:
       remove unneeded distupgrade check
       remove package install and timer enable since that will be handled
         by OS install.

commit edf5d9272e43b1aacb4a4a87efc6696354cced58
Author: sampsone <esampson@suse.com>
Date:   Tue May 21 16:28:32 2024 -0500

    Add support for suse uptime tracker

    The suse uptime tracker package records ever hour a system is up
    as bit mask. This information if present, needs to be collect by suma and
    reported to SCC.
